aten/src/ATen/native/cpu/UpSampleKernel.cpp

```diff
@@ -1244,6 +1282,123 @@ void separable_upsample_generic_Nd_kernel_impl(
       F>(output, temp_input, 2, align_corners, scales);
 }
 
+bool can_use_fast_interpolation(const Tensor& output, const Tensor& input) {
+  auto supported_memory_format = at::MemoryFormat::Contiguous;
+  return input.is_contiguous(supported_memory_format) &&
+      output.is_contiguous(supported_memory_format) &&
+      (input.dtype() == kFloat || input.dtype() == kDouble);
+}
+
+template <typename scalar_t, typename scale_type, class F>
+void upsample_generic_2d_aa_contig(
+    const Tensor& output,
+    const Tensor& input,
+    bool align_corners,
+    const scale_type& scales) {
+  auto input_shape = input.sizes().vec();
+  auto output_shape = output.sizes().vec();
+
+  // fold n and c
+  int64_t channels = input_shape[0] * input_shape[1];
+  int64_t input_height = input_shape[2];
+  int64_t input_width = input_shape[3];
+  int64_t output_height = output_shape[2];
+  int64_t output_width = output_shape[3];
+
+  scalar_t* input_data = input.data_ptr<scalar_t>();
+  scalar_t* output_data = output.data_ptr<scalar_t>();
+
+  using Vec = vec::Vectorized<scalar_t>;
+  using integer_t = vec::int_same_size_t<scalar_t>;
+  using iVec = vec::Vectorized<integer_t>;
+  // for the convience of vectorization, use integer of the same size of scalar_t,
+  // need to make sure doesn't overflow
+  TORCH_CHECK(input_height * input_width <= std::numeric_limits<integer_t>::max());
+
+  std::vector<Tensor> indices_weights_h = F::compute_indices_weights_aa(
+      input.scalar_type(), input_height, output_height, align_corners, scales[0]);
+  std::vector<Tensor> indices_weights_w = F::compute_indices_weights_aa(
+      input.scalar_type(), input_width, output_width, align_corners, scales[1]);
+
+  integer_t* idx_min_h = indices_weights_h[0].data_ptr<integer_t>();
+  integer_t* idx_size_h = indices_weights_h[1].data_ptr<integer_t>();
+  integer_t* idx_min_w = indices_weights_w[0].data_ptr<integer_t>();
+  integer_t* idx_size_w = indices_weights_w[1].data_ptr<integer_t>();
+
+  // restore interp_size based on wts shape
+  int64_t interp_size_h = indices_weights_h[3].numel() / output_height;
+  int64_t interp_size_w = indices_weights_w[3].numel() / output_width;
+
+  // leave `wt_h` in {output_height, interp_size_h} since h will use scalar logic
+  // transpose `wt_w` to {interp_size_w, output_width} since w will use vectorized logic
+  scalar_t* wt_h = indices_weights_h[3].data_ptr<scalar_t>();
+  auto trans_wt = indices_weights_w[3].reshape({output_width, interp_size_w}).transpose(0, 1).contiguous();
+  scalar_t* wt_w = trans_wt.data_ptr<scalar_t>();
+
+  // parallel on dimensions of n, c and h, vectorize on w.
```

Comment:
@mingfeima I wonder if we could vectorize inside `interpolate_aa_single_dim` using tensor iterator approach and if this would have similar perfs : https://github.com/pytorch/pytorch/blob/a5b59f1e2bcee35e20aded58ef266c9855b76d83/aten/src/ATen/native/cpu/UpSampleKernel.cpp#L148-L171
What do you think ?

Reply:
I checked on VTune previously and this part takes very small part in the total runtime. I think it is OK to leave it what it as.